src/renderer.c

```diff
@@ -373,27 +390,38 @@ double ren_draw_text(RenSurface *rs, RenFont **fonts, const char *text, size_t l
   double pen_x = x * surface_scale;
   y *= surface_scale;
   int bytes_per_pixel = surface->format->BytesPerPixel;
-  const char* end = text + len;
+  // const char* end = text + len;
   uint8_t* destination_pixels = surface->pixels;
   int clip_end_x = clip.x + clip.w, clip_end_y = clip.y + clip.h;
 
   RenFont* last = NULL;
   double last_pen_x = x;
   bool underline = fonts[0]->style & FONT_STYLE_UNDERLINE;
   bool strikethrough = fonts[0]->style & FONT_STYLE_STRIKETHROUGH;
-
-  while (text < end) {
-    unsigned int codepoint, r, g, b;
-    text = utf8_to_codepoint(text, &codepoint);
+  // convert text in glyphs
+  hb_buffer_t *buf;
+  buf = hb_buffer_create();
+  hb_buffer_set_direction(buf, HB_DIRECTION_LTR);
+  hb_buffer_add_utf8(buf, text, -1, 0, -1);
+
+  RenFont * font = fonts[0]; 
+  hb_shape(font->font, buf, NULL, 0);
```

Comment:
Question, in case that a fallback font is used for some codepoints not available on main font, shouldn't that portion get shaped also with the fallback font instead of using the main font?

Reply:
Good question, as it is implemented now, ligatures are supported only for the main font and so the shaping is only applied to the main one, for the fallaback case I am using the `fb_codepoint` that is computed in `src/renderer.c:413` that is then converted if needed to the correct font index using `FT_Get_Char_Index` at `src/renderer.c:181`. 
I hope the explanation is clear.
It would be possible to shape for all the available fonts but I suspect will be more computational intensive and you could end up with half a ligature of a fallback font and half a normal character form another. 
What do you think?

Reply:
The function shared on https://github.com/pragtical/pragtical/pull/39/commits/ed28c3ae7d079fe28f9cc174291a99db7789353a iterates a string giving you the range of a string (thru the `char** next` out param) and returns the font that has all of its codepoints for rendering:

```c
static RenFont* utf8_iterator(
  RenFont** fonts, const char* start, const char* end, char** next
) {
  if (start == end) return NULL;

  unsigned int codepoint;
  RenFont* current_font = fonts[0];

  bool iterated = false;
  *next = (char*) start;
  char* prev_next = *next;
  while(*next < end) {
    *next = (char*) utf8_to_codepoint(*next, &codepoint);
    if (!FT_Get_Char_Index(current_font->face, codepoint)) {
      if (!iterated) {
        bool codepoint_found = false;
        for (short i = 1; i < FONT_FALLBACK_MAX && fonts[i]; ++i) {
          if (FT_Get_Char_Index(fonts[i]->face, codepoint)) {
            current_font = fonts[i];
            codepoint_found = true;
            break;
          }
        }
        if (!codepoint_found)
          return fonts[0];
      } else {
        // set next to the end of previously iterated text range
        *next = prev_next;
        return current_font;
      }
    }
    // Allow emojis to be rendered with different fonts, this is needed because
    // sometimes a single glyph can be queried, which could be on the first
    // font and also on previously used fallbacks when rendering.
    if (!iterated && (*next - start) > 3) {
      return current_font;
    }
    prev_next = *next;
    iterated = true;
  }

  return current_font;
}
```
This way you can shape string segments using the assigned font (which is the strategy used on that patch). When tested I didn't notice a performance drop but I didn't benchmark the thing :)